### ISSUE TITLE
Extremely slow file handling with large archives in Android

### ISSUE THREAD

**Issue by Jaifroid** [openzim/javascript-libzim#42]
Issue was initially discussed and diagnosed over at https://github.com/kiwix/kiwix-js-windows/issues/343. There, I thought the issue was with Emscripten's WORKERFS. However, using the [test case in this Repo for large file access](https://openzim.github.io/javascript-libzim/tests/test_large_file_access/), and debugging on a Chromium instance on a midrange Samsung Android, the WORKERFS has no problem loading and reading (instantly) a 92GB Wikipedia ZIM from a microSD card. See screenshot below. If we can read bytes from the end of the file nearly instantly, why is javascript-libzim getting such awful performance? Loading Ray Charles into the WASM on Android takes 30 seconds in Samsung Internet (a Chromium browser), and nearly 90 seconds in Chrome. I was unable to load anything larger than 500MB into either instance.

The slowdown is mostly in relation to instantiating the archive. Once the archive has registered, full-text searching is reasonable. This is why I initially thought the issue was to do with passing the ArrayBuffer to the Web Worker, but big file test does this instantly. @mgautierfr would you have any thoughts on what could be going on here?

![image](https://user-images.githubusercontent.com/4304337/211220695-227aff78-6479-45c2-89b0-d70fb3a47480.png)

NB we can't test on Firefox because Firefox on Android attempts to copy picked archives into memory (or possibly an internal file system), and crashes on anything larger than about 2GB.

**Comment by Jaifroid** [comment]
I think I've narrowed it down to whatever lies behind [this code in the bindings](https://github.com/openzim/javascript-libzim/blob/main/libzim_bindings.cpp#L21):

<img alt="image" src="https://user-images.githubusercontent.com/4304337/211222518-2fbc9989-e4c4-4537-8335-04f2e7028758.png">

The app hangs (in Android) for a ridiculously long time even with Ray Charles on this JS invocation of the above code:

<img alt="image" src="https://user-images.githubusercontent.com/4304337/211222607-ea887c79-3a49-4b99-9f10-367ac408e273.png">

**Comment by Jaifroid** [comment]
Perhaps @mgautierfr's analysis of search performance in https://github.com/openzim/libzim/issues/418 is relevant to this issue, in particular the observation that most of the "lost" time is spent on I/O. However, the real bottleneck we're experiencing is in the function that loads the ZIM archive. Unless some of the caching ideas regarding Xapian search have been implemented, just loading the archive should not even initiate Xapian code, right?

**Comment by mgautierfr** [comment]
I don't know how you "system" is architectured. Where the io is happening ? When you want to access libzim or when libzim access the file ?

You speak about fulltext search. fulltext search is made by xapian library itself and I don't know how it is working. But it seems that loading the xapian database need a lot of io (we can humanly see that even locally in cpp). Once the database is loaded, following search are pretty quick. Maybe it is the same problem but augmented by a io a bit slower in JS/SW/WebWorker (even if 90s seem really a lot, even for that)

Your test for access to large file seems to get only one bytes (at different offsets). How it is behaving when you try to read several bytes (few KB, few MB, few hundreds of MB) ? How it is behaving when you try to access at random offsets (not only increasing offsets) ?

**Comment by mgautierfr** [comment]
Can you have a trace of all io made by libzim (how many bytes read, at which offsets) ?

**Comment by Jaifroid** [comment]
Good idea -- although I probably can't do that for the WASM (which is a kind of machine assembly code), I can probably do it in the ASM version which is JavaScript bytecode, and so vaguely human-readable and traceable/debuggable.
